src/box/memtx_tx.c

```diff
@@ -345,21 +345,69 @@ memtx_tx_cause_conflict(struct txn *breaker, struct txn *victim)
 	return 0;
 }
 
+/**
+ * Fix read view list if rv_psn of a transaction was decreased.
+ */
+static void
+memtx_tx_handle_rv_psn_decrease(struct txn *txn)
+{
+	/*
+	 * The list of read views must be ordered by rv_psn. If we decrease
+	 * rv_psn in a transaction we probably must move its node to the head
+	 * of the list.
+	 */
+	if (txn->in_read_view_txs.prev == &txm.read_view_txs)
+		return; /* No transaction before */
+	struct txn *prev_txn = rlist_prev_entry(txn, in_read_view_txs);
+	if (prev_txn->rv_psn <= txn->rv_psn)
+		return; /* The order is already correct. */
+	/* Remove from list for a while. */
+	rlist_del(&txn->in_read_view_txs);
+	while (prev_txn->in_read_view_txs.prev != &txm.read_view_txs) {
+		struct txn *scan = rlist_prev_entry(prev_txn, in_read_view_txs);
+		if (scan->rv_psn <= txn->rv_psn)
+			break;
+		prev_txn = scan;
+	}
+	/* Insert before prev_txn. */
+	rlist_add_tail(&prev_txn->in_read_view_txs, &txn->in_read_view_txs);
+}
+
 void
 memtx_tx_handle_conflict(struct txn *breaker, struct txn *victim)
 {
 	assert(breaker->psn != 0);
-	if (victim->status != TXN_INPROGRESS) {
+	if (victim->status != TXN_INPROGRESS &&
+	    victim->status != TXN_IN_READ_VIEW) {
 		/* Was conflicted by somebody else. */
 		return;
 	}
 	if (stailq_empty(&victim->stmts)) {
-		/* Send to read view. */
-		victim->status = TXN_IN_READ_VIEW;
-		victim->rv_psn = breaker->psn;
-		rlist_add_tail(&txm.read_view_txs, &victim->in_read_view_txs);
+		assert((victim->status == TXN_IN_READ_VIEW) ==
+		       (victim->rv_psn != 0));
+		/* Send to read view, perhaps a deeper one. */
+		if (victim->status != TXN_IN_READ_VIEW) {
+			victim->status = TXN_IN_READ_VIEW;
+			victim->rv_psn = breaker->psn;
+			rlist_add_tail(&txm.read_view_txs,
+				       &victim->in_read_view_txs);
```

Comment:
Should not we call `memtx_tx_handle_rv_psn_decrease(victim)` here?

Reply:
Yes, thank we should. Added another test that shows that.